internals/overlord/servstate/manager_test.go

```diff
@@ -1753,72 +1560,363 @@ func (s *S) TestStopRunning(c *C) {
+	s.manager, err = servstate.NewManager(s.st, s.runner, dir, nil, nil, fakeLogManager{})
 	c.Assert(err, IsNil)
-	defer manager.Stop()
 
 	outputPath := filepath.Join(dir, "output")
 	layer := parseLayer(c, 0, "layer", fmt.Sprintf(`
 services:
     wrkdir:
         override: replace
-        command: /bin/sh -c "pwd >%s; sleep %g"
+        command: /bin/sh -c "pwd >%s; sync; touch %s; sleep %g"
         working-dir: %s
-`, outputPath, shortOkayDelay.Seconds()+0.01, dir))
-	err = manager.AppendLayer(layer)
+`, outputPath, s.donePath, shortOkayDelay.Seconds()+0.01, dir))
+	err = s.manager.AppendLayer(layer)
 	c.Assert(err, IsNil)
 
 	chg := s.startServices(c, []string{"wrkdir"}, 1)
 	s.st.Lock()
 	c.Assert(chg.Err(), IsNil)
 	s.st.Unlock()
 
+	// Let's wait until we know the output file is available
+	s.waitForFileToAppear(c)
+
 	output, err := ioutil.ReadFile(outputPath)
 	c.Assert(err, IsNil)
 	c.Check(string(output), Equals, dir+"\n")
 }
+
+// Default service manager setup
+func (s *S) defaultServiceManagerSetup(c *C) {
+	s.dir = c.MkDir()
+	s.log = filepath.Join(s.dir, "log.txt")
+	os.Mkdir(filepath.Join(s.dir, "layers"), 0755)
+	data := fmt.Sprintf(planLayer1, s.log, s.log)
+	err := ioutil.WriteFile(filepath.Join(s.dir, "layers", "001-base.yaml"), []byte(data), 0644)
+	c.Assert(err, IsNil)
+	err = ioutil.WriteFile(filepath.Join(s.dir, "layers", "002-two.yaml"), []byte(planLayer2), 0644)
+	c.Assert(err, IsNil)
+
+	s.manager, err = servstate.NewManager(s.st, s.runner, s.dir, s.logOutput, testRestarter{s.stopDaemon}, fakeLogManager{})
+	c.Assert(err, IsNil)
+}
+
+// Return on timeout or when the file appears. This is used to determine
+// when the expected service output is actually available, not when the
+// service starts to run.
+func (s *S) waitForFileToAppear(c *C) {
+	timeout := time.After(time.Duration(time.Second * 10))
+found:
+	for {
+		select {
+		case <-timeout:
+			c.Fatal("timeout waiting for service output")
+		default:
+			stat, err := os.Stat(s.donePath)
+			if err == nil && stat.Mode().IsRegular() {
+				// Delete it so we can reuse this feature
+				// in the same test.
+				os.Remove(s.donePath)
+				break found
+			}
+			time.Sleep(time.Millisecond)
+		}
+	}
+}
+
+// Make sure services are all stopped before the next test starts.
+func (s *S) stopRunningServices(c *C) {
+	taskSet, err := servstate.StopRunning(s.st, s.manager)
+	if err != nil {
+		c.Fatal("error stopping running test services")
+		return
+	}
+	if taskSet == nil {
+		return
+	}
+
+	// One change to stop them all.
+	s.st.Lock()
+	chg := s.st.NewChange("stop", "Stop all running services")
+	chg.AddAll(taskSet)
+	s.st.EnsureBefore(0) // start operation right away
+	s.st.Unlock()
+
+	// Wait for a limited amount of time for them to stop.
+	timeout := time.After(time.Duration(time.Second * 10))
+stop:
+	for {
+		s.runner.Ensure()
+		s.runner.Wait()
+
+		// Exit loop if change is complete
+		select {
+		case <-chg.Ready():
+			break stop
```

Comment:
Could again avoid the labelled break with a `return` here.

Reply:
Done! Thanks.